File: backend/studi0trace/engines/vexel/upsample.py
"""Stage 0b: trace a small input at twice its size when it has thin features.

At 128 px the anti-aliasing model that places edges has two pixels of
evidence per edge, and a feature a couple of pixels wide is placed 0.1–1.9 px
off its truth. Tracing a 2× Lanczos upsample instead and letting the viewBox
carry the scale cut those errors by half to three quarters on every corpus
item with a region under 2.4 px wide (thin-mark 1.73 → 0.76, sticker 0.41 →
0.12, blobs 0.13 → 0.075) — and made every item made of large shapes worse,
because the resampler's ringing and its own anti-aliasing are read as edge
position (cutout 0.013 → 0.121). So the upsample is selected, never applied
blind: the direct trace runs first, and only a label map with a region
thinner than THIN_WIDTH px (width = 2·area/perimeter) sends the image back
through at 2×. The rule needs no renderer, so both engines apply it alike.

The kernel is Lanczos-3 at exactly 2×, which makes the tap weights two fixed
sets of six numbers; they are written here as literals, and the Rust twin
carries the same literals, so the two upsamples agree to the byte
(`tools/diffcheck.py upsample`).
"""
from __future__ import annotations

import numpy as np

UPSAMPLE_MAX_SIDE = 192   # px; larger inputs have the evidence they need
THIN_WIDTH = 2.2          # px; a region narrower than this (2·area/perimeter) wants the upsample
MIN_AREA = 6              # px²; specks below this are not features
# The 2× pass runs with the user's parameters as they are. Scaling min_region
# and curve_tolerance to source pixels was tried and gave back most of the
# outline gain (blobs 0.075 → 0.119) without curing the one item whose
# gradient bands split at 2× (sticker); that item is not selected instead.

# Lanczos-3 weights for an output pixel centred a quarter pixel before (even
# outputs) and after (odd outputs) input pixel i, over taps i-2 … i+3.
_W_EVEN = (-0.06850269446295365, 0.273025020812246, 0.8994068413745139, -0.13426528114738912, 0.030336113423582882, 0.0)
_W_ODD = (0.030112285361897653, -0.1332746355359615, 0.8927707740853273, 0.2710105682570789, -0.06799726302855182, 0.0073782708602093345)


def _pass(a: np.ndarray) -> np.ndarray:
    """Double the first axis. `a` is float64 (n, …)."""
    n = a.shape[0]
    out = np.empty((2 * n, *a.shape[1:]), dtype=np.float64)
    idx = np.arange(n)
    for parity, weights in ((0, _W_EVEN), (1, _W_ODD)):
        acc = np.zeros_like(a)
        for k, w in zip(range(-2, 4), weights):
            src = np.clip(idx + k, 0, n - 1)
            acc = acc + w * a[src]       # one tap at a time, in tap order: the Rust adds in the same order
        out[parity::2] = acc
    return out


def upsample2x(rgba: np.ndarray) -> np.ndarray:
    """(H, W, 4) uint8 → (2H, 2W, 4) uint8, Lanczos-3, channels straight."""
    a = rgba.astype(np.float64)
    a = _pass(a)                          # rows
    a = _pass(a.transpose(1, 0, 2)).transpose(1, 0, 2)  # columns
    return np.clip(np.floor(a + 0.5), 0.0, 255.0).astype(np.uint8)


def thinnest_region(labels: np.ndarray, min_area: int = MIN_AREA) -> float:
    """The smallest 2·area/perimeter over regions (label ≠ 0) of at least
    `min_area` pixels; perimeter counts 4-neighbour label changes, frame edges
    excluded, so a region on the frame reads wider, never thinner. inf if none."""
    best = float("inf")
    for lab in np.unique(labels):
        if lab == 0:
            continue
        m = labels == lab
        area = int(m.sum())
        if area < min_area:
            continue
        per = int((m[:, 1:] != m[:, :-1]).sum()) + int((m[1:, :] != m[:-1, :]).sum())
        best = min(best, 2.0 * area / max(per, 1))
    return best


def wants_upsample(labels: np.ndarray, height: int, width: int) -> bool:
    return max(height, width) <= UPSAMPLE_MAX_SIDE and thinnest_region(labels) < THIN_WIDTH


def halve(svg: str, width: int, height: int) -> str:
    """The SVG of the 2× trace, drawn at the original size: the root viewBox is
    the original canvas and everything but the defs sits in a half-scale group.
    Gradients (userSpaceOnUse), filters and stroke widths follow the element's
    user space, so they scale with it."""
    import re

    m = re.match(r"(<svg[^>]*>)(<defs>.*?</defs>)?(.*)</svg>$", svg, re.S)
    if not m:
        return svg
    root = re.sub(r'viewBox="[^"]*"', f'viewBox="0 0 {width} {height}"', m.group(1), count=1)
    return f'{root}{m.group(2) or ""}<g transform="scale(0.5)">{m.group(3)}</g></svg>'

File: backend/vexel-rs/src/upsample.rs
//! Stage 0b: trace a small input at twice its size when it has thin features.
//! Mirrors `upsample.py`, literal for literal, so the two upsamples agree to the byte.

use crate::core::labels::Labels;

pub const UPSAMPLE_MAX_SIDE: usize = 192;
pub const THIN_WIDTH: f64 = 2.2;
pub const MIN_AREA: usize = 6;

const W_EVEN: [f64; 6] = [-0.06850269446295365, 0.273025020812246, 0.8994068413745139, -0.13426528114738912, 0.030336113423582882, 0.0];
const W_ODD: [f64; 6] = [0.030112285361897653, -0.1332746355359615, 0.8927707740853273, 0.2710105682570789, -0.06799726302855182, 0.0073782708602093345];

/// Double the row count of an (n, m) f64 image with `c` channels: out[(2i+parity), j, ch].
fn pass_rows(a: &[f64], n: usize, m: usize, c: usize) -> Vec<f64> {
    let mut out = vec![0.0f64; 2 * n * m * c];
    for (parity, weights) in [(0usize, &W_EVEN), (1usize, &W_ODD)] {
        for i in 0..n {
            for j in 0..m {
                for ch in 0..c {
                    let mut acc = 0.0f64;
                    for (t, w) in weights.iter().enumerate() {
                        let k = i as i64 + t as i64 - 2;
                        let src = k.clamp(0, n as i64 - 1) as usize;
                        acc += w * a[(src * m + j) * c + ch];
                    }
                    out[((2 * i + parity) * m + j) * c + ch] = acc;
                }
            }
        }
    }
    out
}

fn transpose(a: &[f64], n: usize, m: usize, c: usize) -> Vec<f64> {
    let mut out = vec![0.0f64; n * m * c];
    for i in 0..n {
        for j in 0..m {
            for ch in 0..c {
                out[(j * n + i) * c + ch] = a[(i * m + j) * c + ch];
            }
        }
    }
    out
}

/// (H, W, 4) u8 → (2H, 2W, 4) u8, Lanczos-3, channels straight. See the Python `upsample2x`.
pub fn upsample2x(rgba: &[u8], h: usize, w: usize) -> Vec<u8> {
    let a: Vec<f64> = rgba.iter().map(|v| *v as f64).collect();
    let rows = pass_rows(&a, h, w, 4); // (2h, w)
    let t = transpose(&rows, 2 * h, w, 4); // (w, 2h)
    let cols = pass_rows(&t, w, 2 * h, 4); // (2w, 2h)
    let back = transpose(&cols, 2 * w, 2 * h, 4); // (2h, 2w)
    back.iter().map(|v| (v + 0.5).floor().clamp(0.0, 255.0) as u8).collect()
}

/// The smallest 2·area/perimeter over regions of at least MIN_AREA pixels; see the Python.
pub fn thinnest_region(l: &Labels) -> f64 {
    let (h, w) = (l.h, l.w);
    let mut ids: Vec<i32> = l.data.iter().copied().filter(|v| *v != 0).collect();
    ids.sort_unstable();
    ids.dedup();
    let mut best = f64::INFINITY;
    for lab in ids {
        let mut area = 0usize;
        let mut per = 0usize;
        for y in 0..h {
            for x in 0..w {
                let here = l.data[y * w + x] == lab;
                if here {
                    area += 1;
                }
                if x + 1 < w && here != (l.data[y * w + x + 1] == lab) {
                    per += 1;
                }
                if y + 1 < h && here != (l.data[(y + 1) * w + x] == lab) {
                    per += 1;
                }
            }
        }
        if area < MIN_AREA {
            continue;
        }
        best = best.min(2.0 * area as f64 / per.max(1) as f64);
    }
    best
}

pub fn wants_upsample(l: &Labels, height: usize, width: usize) -> bool {
    height.max(width) <= UPSAMPLE_MAX_SIDE && thinnest_region(l) < THIN_WIDTH
}

/// The SVG of the 2× trace drawn at the original size. See the Python `halve`.
pub fn halve(svg: &str, width: usize, height: usize) -> String {
    let Some(root_end) = svg.find('>') else { return svg.to_string() };
    let root = &svg[..=root_end];
    let rest = &svg[root_end + 1..];
    let Some(body) = rest.strip_suffix("</svg>") else { return svg.to_string() };
    let root = match (root.find("viewBox=\""), root.find("viewBox=\"").and_then(|s| root[s + 9..].find('"').map(|e| s + 9 + e))) {
        (Some(s), Some(e)) => format!("{}viewBox=\"0 0 {} {}\"{}", &root[..s], width, height, &root[e + 1..]),
        _ => root.to_string(),
    };
    let (defs, elements) = if body.starts_with("<defs>") {
        match body.find("</defs>") {
            Some(e) => (&body[..e + 7], &body[e + 7..]),
            None => ("", body),
        }
    } else {
        ("", body)
    };
    format!("{}{}<g transform=\"scale(0.5)\">{}</g></svg>", root, defs, elements)
}
